### PR TITLE
Assignment 4 - Matthew Powers - Spencer West

### PR DESCRIPTION
**Pull request by Matthew-Powers**
@SaaWest @Matthew-Powers

### CODE REVIEW

**Comment by SergeiAshikhmin** [comment]
I will review this

**Review by SergeiAshikhmin — COMMENTED** [review]
Overall the code looks good. Good job 👍
I feel like you are not done yet with this assignment but what you have right now is good.

Tasks
Define a full name record (first, last, middle) handling optional and null appropriately. ✔
Provide a comment on the full name record on why you selected to define a value or a reference type and ✔
Provide a comment on the full name record on why or why not the type is immutable. ✔
Define an IEntity interface: ✔
Add an Id property of type Guid that is an init-only setter. ✔
Add a Name property that is string. ✔
Define an abstract base class that implements IEntity - appropriately choosing to implement the interface explicitly or implicitly. ✔
Do not implement the Name property in this abstract class. ✔
Do force any derived classes to provide an implementation for Name. ✔
Define book, student, and employee records - all with entity capabilities. ❌
Choose a thoughtful approach for the value of Name for each type of entity. In all cases, there should be no backing field for Name (not even one generated by an automatically implemented property) - consider a calculated property. ❌
Provide a comment on each interface member in each entity explaining why you implemented it implicitly or explicitly. ❌
You should consider the relationship between Student and Employee and refactor the common code shared between them. ❌
Given the properties on the full name record, you should consider which entities it makes sense to use it. ❌
Test that the equality behavior on these entities behaves as expected. ❌
Provide a comment on each interface method explaining why you implemented it implicitly or explicitly.
Unit test the Storage class using various entities to ensure that there are no bugs.
Code review (using TODO: comments if there are any issues - which there may not be) all logger-related classes.)
Pay special attention to the CreateLogger methods throughout.

Fundamentals
Be sure you enable:
Enable Nullability for all projects ✔
Set LangVersion and the TargetFramework to the latest released versions available (preview versions optional) ✔
Ensure that you turn on code analysis for all projects(EnableNETAnalyzers) ✔
Ensure that you turn on EnforceCodeStyleInBuild ✔
Ensure that you turn on Nullability (Nullable) ✔
Ensure there are no errors or warnings (including code analysis warnings) ✔
All of the above should be unit tested ❌
Choose simplicity over complexity ✔

File: Logger.Tests/Logger.Tests.csproj [hunk 1]
```diff
@@ -7,14 +7,18 @@
     <LangVersion>12.0</LangVersion>
```
Comment (robertogarcia11) [inline Logger.Tests/Logger.Tests.csproj:7]:
```suggestion
    <LangVersion>12.0</LangVersion>
    <TargetFramework>net8.0</TargetFramework>

```

**Comment by robertogarcia11** [comment]
Good start, but I agree, seems like you are not done yet!

Tasks
Define a full name record (first, last, middle) handling optional and null appropriately. ✔
Provide a comment on the full name record on why you selected to define a value or a reference type and ✔
Provide a comment on the full name record on why or why not the type is immutable. ✔
Define an IEntity interface: ✔
Add an Id property of type Guid that is an init-only setter. ✔
Add a Name property that is string. ✔
Define an abstract base class that implements IEntity - appropriately choosing to implement the interface explicitly or implicitly. ✔
Do not implement the Name property in this abstract class. ✔
Do force any derived classes to provide an implementation for Name. ✔
Define book, student, and employee records - all with entity capabilities. ❌
Choose a thoughtful approach for the value of Name for each type of entity. In all cases, there should be no backing field for Name (not even one generated by an automatically implemented property) - consider a calculated property. ❌
Provide a comment on each interface member in each entity explaining why you implemented it implicitly or explicitly. ❌
You should consider the relationship between Student and Employee and refactor the common code shared between them. ❌
Given the properties on the full name record, you should consider which entities it makes sense to use it. ❌
Test that the equality behavior on these entities behaves as expected. ❌
Provide a comment on each interface method explaining why you implemented it implicitly or explicitly.
Unit test the Storage class using various entities to ensure that there are no bugs.
Code review (using TODO: comments if there are any issues - which there may not be) all logger-related classes.)
Pay special attention to the CreateLogger methods throughout.

Fundamentals
Be sure you enable:
Enable Nullability for all projects ✔
Set LangVersion and the TargetFramework to the latest released versions available (preview versions optional) ✔/❌  [missing the TargetFramework  for the tests proj file, unless I missed it]
Ensure that you turn on code analysis for all projects(EnableNETAnalyzers) ✔
Ensure that you turn on EnforceCodeStyleInBuild ✔
Ensure that you turn on Nullability (Nullable) ✔
Ensure there are no errors or warnings (including code analysis warnings) ✔
All of the above should be unit tested ❌ [needs more testing]
Choose simplicity over complexity ✔

File: Logger/ILogConfiguration.cs [hunk 2]
```diff
@@ -3,5 +3,4 @@
 public interface ILoggerConfiguration
```
Comment (EthanFlorko) [inline Logger/ILogConfiguration.cs:3]: Add comments for the interface to explain implicitly or explicitly.

File: Logger/IEntity.cs [hunk 3]
```diff
@@ -1,7 +1,6 @@
 ﻿namespace Logger;
 public interface IEntity
```
Comment (EthanFlorko) [inline Logger/IEntity.cs:2]: Add comments for the interface to explain implicitly or explicitly.

**Comment by EthanFlorko** [comment]
Just missing a few tasks listed below, but it was a good start. Don't give up!

Tasks
Define a full name record (first, last, middle) handling optional and null appropriately. ✔
Provide a comment on the full name record on why you selected to define a value or a reference type and ✔
Provide a comment on the full name record on why or why not the type is immutable. ✔
Define an IEntity interface: ✔
Add an Id property of type Guid that is an init-only setter. ✔
Add a Name property that is string. ✔
Define an abstract base class that implements IEntity - appropriately choosing to implement the interface explicitly or implicitly. ✔
Do not implement the Name property in this abstract class. ✔
Do force any derived classes to provide an implementation for Name. ✔
Define book, student, and employee records - all with entity capabilities. ❌
Choose a thoughtful approach for the value of Name for each type of entity. In all cases, there should be no backing field for Name (not even one generated by an automatically implemented property) - consider a calculated property. ❌
Provide a comment on each interface member in each entity explaining why you implemented it implicitly or explicitly. ❌
You should consider the relationship between Student and Employee and refactor the common code shared between them. ❌
Given the properties on the full name record, you should consider which entities it makes sense to use it. ❌
Test that the equality behavior on these entities behaves as expected. ❌
Provide a comment on each interface method explaining why you implemented it implicitly or explicitly.
Unit test the Storage class using various entities to ensure that there are no bugs.
Code review (using TODO: comments if there are any issues - which there may not be) all logger-related classes.)
Pay special attention to the CreateLogger methods throughout.

Fundamentals
Be sure you enable:
Enable Nullability for all projects ✔
Set LangVersion and the TargetFramework to the latest released versions available (preview versions optional) ✔ [missing the TargetFramework for the tests proj file, unless I missed it]
Ensure that you turn on code analysis for all projects(EnableNETAnalyzers) ✔
Ensure that you turn on EnforceCodeStyleInBuild ✔
Ensure that you turn on Nullability (Nullable) ✔
Ensure there are no errors or warnings (including code analysis warnings) ✔
All of the above should be unit tested ❌ [needs more testing]
Choose simplicity over complexity ✔

File: Logger/FullName.cs [hunk 4]
```diff
@@ -0,0 +1,23 @@
+﻿namespace Logger;
+
+
+public record FullName
```
[the remark this answers is not in the document]
Comment (SaaWest) [reply Logger/FullName.cs:4]: will change it

File: Logger/FullName.cs [hunk 5]
```diff
@@ -0,0 +1,23 @@
...
+    //Records are immutable and so FullName should reflect that as first, middle, and last names do
+    //not change in real life unless some legal entity supercedes that.
+    public FullName(string firstName, string lastName, string? middleName = null)
+    {
+        FirstName = string.IsNullOrEmpty(firstName) ? throw new ArgumentNullException(nameof(firstName)) : firstName;
+        LastName = string.IsNullOrEmpty(lastName) ? throw new ArgumentNullException(nameof(lastName)) : lastName;
+        MiddleName = middleName;
+    }
+
+    public string FirstName { get; set; }
```
[the remark this answers is not in the document]
Comment (SaaWest) [reply Logger/FullName.cs:17]: An oversight that will be corrected

File: Logger/BookRecord.cs [hunk 6]
```diff
@@ -0,0 +1,15 @@
...
+{
+    public BookRecord(string bookName, string authorName)
+    {
+        BookName = bookName ?? throw new ArgumentNullException(nameof(bookName));
+        AuthorName = authorName ?? throw new ArgumentNullException(nameof(authorName));
+
+    }
+    public string BookName { get; set; }
+    public string AuthorName { get; set; }
+    public Guid ID { init => throw new NotImplementedException(); }
```
[the remark this answers is not in the document]
Comment (SaaWest) [reply Logger/BookRecord.cs:13]: will correct these mistakes

File: Logger/EmployeeRecord.cs [hunk 7]
```diff
@@ -0,0 +1,7 @@
+﻿namespace Logger;
+
+public record EmployeeRecord : IEntity
+{
+    public Guid ID { init => throw new NotImplementedException(); }
```
[the remark this answers is not in the document]
Comment (SaaWest) [reply Logger/EmployeeRecord.cs:5]: will correct these mistakes

File: Logger/StudentRecord.cs [hunk 8]
```diff
@@ -0,0 +1,7 @@
+﻿namespace Logger;
+
+public record StudentRecord : IEntity
+{
+    public Guid ID { init => throw new NotImplementedException(); }
```
[the remark this answers is not in the document]
Comment (SaaWest) [reply Logger/StudentRecord.cs:5]: will correct these mistakes

File: Logger/ILogConfiguration.cs (hunk 2 above)
Comment (SaaWest) [reply Logger/ILogConfiguration.cs:3]: will do

File: .devcontainer/devcontainer.json
Comment (BenjaminMichaelis) [inline .devcontainer/devcontainer.json:1]: Your git branch is in a bad state, that's why it's showing all these additions that are already in the base branch

File: Logger.Tests/Logger.Tests.csproj [hunk 9]
```diff
@@ -5,16 +5,22 @@
...
+    <EnforceCodeStyleInBuild>true</EnforceCodeStyleInBuild>
+    <TreatWarningsAsErrors>true</TreatWarningsAsErrors>
   </PropertyGroup>
 
   <ItemGroup>
     <PackageReference Include="coverlet.collector" Version="6.0.0">
       <PrivateAssets>all</PrivateAssets>
       <IncludeAssets>runtime; build; native; contentfiles; analyzers; buildtransitive</IncludeAssets>
     </PackageReference>
+    <PackageReference Include="DocuSign.eSign.dll" Version="6.8.0" />
```
Comment (BenjaminMichaelis) [inline Logger.Tests/Logger.Tests.csproj:21]: What is this being used for?

File: Logger.Tests/StorageTests.cs [hunk 10]
```diff
@@ -0,0 +1,86 @@
...
+namespace Logger.Tests;
+public class StorageTests
+{
+    [Fact]
+    public void Storage_AddBook_Entity()
+    {
+        var book = new BookRecord("Do Robots Dream of Electric Sheep", "Philip K. Dick");
+        var storage = new Storage();
+        storage.Add(book);
+        Assert.True(storage.Contains(book));;
```
Comment (BenjaminMichaelis) [inline Logger.Tests/StorageTests.cs:12]: Why not use `Assert.Contains()`?

File: Logger.Tests/StorageTests.cs [hunk 11]
```diff
@@ -0,0 +1,86 @@
...
+    }
+    [Fact]
+    public void Storage_RemoveBook_Entity() 
+    {
+        var book = new BookRecord("Do Robots Dream of Electric Sheep", "Philip K. Dick");
+        var storage = new Storage();
+        storage.Add(book);
+        Assert.True(storage.Contains(book));
+        storage.Remove(book);
+        Assert.False(storage.Contains(book));
```
Comment (BenjaminMichaelis) [inline Logger.Tests/StorageTests.cs:23]: Why not using `Assert.DoesNotContain()`

File: Logger/BaseEntity.cs [hunk 12]
```diff
@@ -0,0 +1,11 @@
+﻿
+
+namespace Logger;
+
+public abstract record class BaseEntity : IEntity
+{
+    // Both Name and ID are implicit as BaseEntity 'is a' relationship with IEntity. It would not hurt to make ID explicit but by convention of the relationship implicit is chosen but because name
+    // presumably used in multiple record classes name is chosen as an abstract implicit. 
+    public abstract string Name { get; }
+    public Guid ID { init => throw new NotImplementedException(); }
```
Comment (BenjaminMichaelis) [inline Logger/BaseEntity.cs:10]: What is the need of this Exception? Where is the getter?

**Pull request closed without merging**